### ISSUE TITLE
[BUG] [v2]: training loss much lower than test and validation loss

### ISSUE THREAD

**Issue by davidegraff** [chemprop/chemprop#607]
Whenever I train a v2 model with a small dataset and validate/test on the same dataset, the training loss is significantly lower than what the training progress bar would have me expect.

**Example**
```python
import lightning as L
import pandas as pd
from torch.utils.data import DataLoader

from chemprop import nn, models, featurizers
from chemprop.data import MoleculeDatapoint, MoleculeDataset, collate_batch

df = pd.read_csv("tests/data/regression/mol.csv")
smis = df["smiles"].to_list()
Y = df["lipo"].to_numpy().reshape(-1, 1)

batch_size = 50
datapoints = [MoleculeDatapoint.from_smi(smi, y) for smi, y in zip(smis, Y)]
dset = MoleculeDataset(datapoints, featurizer)
dset.normalize_targets()
dataloader = DataLoader(dset, batch_size, shuffle=True, collate_fn=collate_batch)

mp = nn.BondMessagePassing()
agg = nn.SumAggregation()
ffn = nn.RegressionFFN()
model = models.MPNN(mp, agg, ffn, True)

trainer = L.Trainer(max_epochs=100, overfit_batches=2, log_every_n_steps=2)
trainer.fit(model, dataloader, dataloader)
trainer.test(model, dataloader)
```

Over the course of training, the training loss approaches 0 while the validation loss gets lower but eventually stalls around 1e-2. Even more concerning is the test loss, which is reported as being around 0.4-0.5. I'm struggling to understand why this is the case. Note that while `batch_norm=True` does introduce slight differences between training and evaluation modes, I would not expect this large of a difference, especially when using the same dataset for all three scenarios. I might be missing something obvious here, but I can't seem to figure out what it is.

**Comment by hwpang** [comment]
I looked into this and recorded my observation here. I turn the shuffle False in the data loader for now for debugging.

```
dataloader = DataLoader(dset, batch_size, shuffle=False, collate_fn=collate_batch)
```

I found that if I turn off the batch norm, the training loss and validation loss becomes within the same order of magnitude, although still not exactly the same (`train_loss=0.00218, val_loss: 0.00169`).

If I turn on the batch norm, the training loss becomes much smaller than the validation loss (`train_loss=1.98e-11, val_loss=0.010`)

**Comment by davidegraff** [comment]
good point re: batch norm. Do you still observe the same phenomenon when setting `overfit_batches=1`? Also the stored statistics of batch norm use the biased standard deviation during training and the _unbiased_ standard deviation during evaluation, so I guess that's one more source of difference from `train -> eval`. But this still doesn't solve for why the test loss is so much lower. It should be equivalent to the validation loss, as both of these use the same model configuration, i.e., `model.eval()`.

**Comment by hwpang** [comment]
### Differences between val and test

The difference between test loss and val loss seems to be related to `shuffle=True`:

The test loss is different from val loss and train loss when the shuffle is set to True (`train_loss=2.01e-11, val_loss=0.0054, test/mse=0.2895137071609497`)

When I set shuffle to False, the test loss is the same as the val loss. (`train_loss=2.32e-11, val_loss=0.00592, test/mse=0.005923879332840443)

The results above are similar with or without batch norm.

### Differences between train and val

If I set `overfit_batches=1.0, shuffle=False, batch_norm=True`, I observe the same problem where the training loss is much smaller than the val and test loss (`train_loss=1.02e-11, val_loss=0.00497, test/mse=0.0049730525352060795`). 

Note that I use `overfit_batches=1.0` instead of `overfit_batches=1` because I assume that we want to overfit the whole training set but not just one batch. This is from their documentation:

```
overfit_batches: Overfit a fraction of training/validation data (float) or a set number of batches (int).
                Default: ``0.0``.
```

**Comment by davidegraff** [comment]
hmm that's odd re: `shuffle` impacting performance. There should be no order-dependence on predictions/performance. Also, I did mean `overfit_batches=1`. This is so that the fitted statistics of the batch norm layer are _perfectly_ optimized for the single training batch, minimizing difference between train/val/test.

**Comment by KnathanM** [comment]
I think the culprit here is that the running estimates of the mean and variance of the input tensors are initialized with a mean of 0s and variance of 1s and then updated with a momentum (default 0.1). While the training loop uses the exact batch statistics, the validation loop uses the running estimates which are biased towards 0 and 1. So while the model weights are trained based on inputs that are normalized using the exact batch statistics, the validation loop uses inputs that are normalized differently and we should expect the error to be higher. 

As evidence for this, I found that using `self.bn = nn.BatchNorm1d(self.message_passing.output_dim, track_running_stats=False)` instead of `self.bn = nn.BatchNorm1d(self.message_passing.output_dim)` in `class MPNN` makes the big difference between the training loss and val loss go away. `track_running_stats=False` lets the val loop also use the exact batch statistics. 

My code. I use a large hidden dim for the message passing because a large dim size for batch normalization makes the difference between train and val loss more apparent. 
```
import lightning as L
import numpy as np
import pandas as pd
from torch.utils.data import DataLoader

from chemprop import nn, models, featurizers
from chemprop.data import MoleculeDatapoint, MoleculeDataset, collate_batch

df = pd.read_csv("tests/data/regression/mol/mol.csv")
smis = df["smiles"].to_list()
Y = df["lipo"].to_numpy().reshape(-1, 1)

n_dp = 10
dp = [MoleculeDatapoint.from_smi(smi, y) for smi, y in zip(smis[:n_dp], Y)]
dset = MoleculeDataset(dp)
dl = DataLoader(dset, batch_size=n_dp, collate_fn=collate_batch)

dim = 2000 
mp = nn.BondMessagePassing(d_h=dim)
agg = nn.SumAggregation()
ffn = nn.RegressionFFN(input_dim=dim)
model = models.MPNN(mp, agg, ffn, batch_norm=True)

trainer = L.Trainer(max_epochs=10)
trainer.fit(model, dl, dl)
```

I expect this is a feature and not a bug of batch normalization because this helps the model not overfit to the training data. In any event, batch normalization is turned off by default and it is up to the user to see whether it helps, so I will close this issue.
